[feat] Send API message whenever a comment is added or deleted

lpagliari opened the pull request:
Send an API message `comments_set_changed` when the set of comments on pad changes. This includes:

- adding a comment to a text;
- deleting a comment from text;
- deleting text which has a comment;
- pasting text which has comments;
- UNDO and REDO operations that add/remove comment on text;

Implement part of [card #1828](https://trello.com/c/pp2UnQ1B/1828).

lpagliari committed:
[feat] Send API message whenever a comment is added or deleted

Send an API message `comments_set_changed` when the set of comments on
pad changes. This includes:

- adding a comment to a text;
- deleting a comment from text;
- deleting text which has a comment;
- pasting text which has comments;
- UNDO and REDO operations that add/remove comment on text;

Implement part of https://trello.com/c/pp2UnQ1B/1828.

joassouza commented on static/js/commentsSetChangeHandler.js:
```diff
@@ -0,0 +1,55 @@
+// a line with a comment was edited, but we need to check if a comment was added
+// or removed from text.
+// Ignore changes made by other users -- we are only interested on changes of
+//  current user
```
ficou um espacinho aqui

joassouza commented on static/js/commentsSetChangeHandler.js:
```diff
@@ -0,0 +1,55 @@
+// or removed from text.
+// Ignore changes made by other users -- we are only interested on changes of
+//  current user
+commentsSetChangeHandler.prototype._handlePossibleCommentsAddedOrRemovedFromTect = function() {
```
`Tect`

joassouza commented on static/js/commentsSetChangeHandler.js:
```diff
@@ -0,0 +1,55 @@
+    this.commentAddedOrRemoved();
+  }
+
+  this.lastCommentIdsSent = currentCommentIds;
```
o que você acha de ao invés de `padHasMyChanges` a gente colocasse algo `thisUserChangedPad`? (btw, é até o nome da função). Achei que ficou meio complicado entender

lpagliari replied:
não posso ter uma função e uma flag com mesmo nome, uma sobrescreveria a outra

lpagliari replied:
vou tentar pensar em algo

joassouza commented on static/js/commentsSetChangeHandler.js:
```diff
@@ -0,0 +1,55 @@
+  this.api.triggerCommentsSetChanged();
+
+  // update current comment ids
+  this.lastCommentIdsSent = this.commentDataManager.getCommentIdsStillOnText();
```
acho que a gente não precisa disso aqui já que a gente faz o mesmo aqui https://github.com/storytouch/ep_comments/pull/58/files#diff-500ef4438a363f450743d4252b99f12eR44
que será chamada logo depois disso

lpagliari replied:
essa função é chamada em outros lugares, não dá para garantir que sempre que chamar uma vai chamar a outra

joassouza commented on static/js/index.js:
```diff
@@ -510,3 +520,8 @@ exports.collectContentPre = function(hook, context){
   copyPasteEvents.handlePaste(context.e);
 }
+
+exports.handleClientMessage_ACCEPT_COMMIT = function(hook, context) {
```
```
_etherpad: collab_client.js has a pretty extensive list of message types, if you want to take a look._
é claro que não tem as mensagens separadas em um lugar só, com lindos comentários
```
o que vc acha de adicionar um comentário descrevendo quando esse hook é chamado? 
Como é um dos hooks que a gente usa bem pouco no código, talvez seja mais fácil ler o comentário do que mergulhar no core do Etherpad para entender.

lpagliari replied:
boa!

joassouza commented on static/js/index.js:
```diff
@@ -510,3 +520,8 @@ exports.collectContentPre = function(hook, context){
   copyPasteEvents.handlePaste(context.e);
 }
+
+exports.handleClientMessage_ACCEPT_COMMIT = function(hook, context) {
```
dúvida: a gente tem um objeto `hooks` com alguns hooks e outros estão fora desse objeto. Sabe porquê 🤔 ? Por causa disso a gente tem `var thisPlugin = pad.plugins.ep_comments_page;` 2 vezes

lpagliari replied:
acho que eram legados, não foram movidos por preguiça nossa mesmo

lpagliari committed:
[refactor] Address code review suggestions (#58)

lpagliari committed:
[refactor] Backend code organization

- extract `api.js`;
- ensure API returns comment/reply author and creator;
- move `utils.js` used only by backend tests into
`static/tests/backend/` folder;

lpagliari committed:
[feat] Send comment ids on "new data" API event

We need to send the ids of current comments on text because the plugin
API only provides the full set of comment data, not only the set of
comments that are actually still on text. This means that if some text
with a comment is removed from the pad, the plugin API will still return
that comment.

To allow merging the comments data with the list of comments still on
text, we provide the comment ids on the 'comments_set_changed' message
and let consumers merge that with the data from the plugin API on `GET
/p/:pad/:rev?/comments`.
